pkg/middleware/auth_test.go

```diff
@@ -150,6 +155,190 @@ func TestAuth_Middleware(t *testing.T) {
 	}
 }
 
+func TestRoleAppPluginAuth(t *testing.T) {
+	t.Run("Verify user's role when requesting app route which requires role", func(t *testing.T) {
+		appSubURL := setting.AppSubUrl
+		setting.AppSubUrl = "/grafana/"
+		t.Cleanup(func() {
+			setting.AppSubUrl = appSubURL
+		})
+
+		tcs := []struct {
+			roleRequired org.RoleType
+			role         org.RoleType
+			signedIn     bool
+			expStatus    int
+			expBody      string
+			expLocation  string
+		}{
+			{signedIn: false, roleRequired: org.RoleViewer, role: org.RoleAdmin, expStatus: http.StatusFound, expBody: "<a href=\"/grafana/login\">Found</a>.\n\n", expLocation: "/grafana/login"},
+			{signedIn: false, roleRequired: org.RoleAdmin, role: org.RoleAdmin, expStatus: http.StatusFound, expBody: "<a href=\"/grafana/login\">Found</a>.\n\n", expLocation: "/grafana/login"},
+			{signedIn: true, roleRequired: org.RoleViewer, role: org.RoleAdmin, expStatus: http.StatusOK, expBody: ""},
+			{signedIn: true, roleRequired: org.RoleAdmin, role: org.RoleAdmin, expStatus: http.StatusOK, expBody: ""},
+			{signedIn: true, roleRequired: org.RoleAdmin, role: org.RoleViewer, expStatus: http.StatusFound, expBody: "<a href=\"/grafana/\">Found</a>.\n\n", expLocation: "/grafana/"},
+			{signedIn: true, roleRequired: "", role: org.RoleViewer, expStatus: http.StatusOK, expBody: ""},
+			{signedIn: true, roleRequired: org.RoleEditor, role: "", expStatus: http.StatusFound, expBody: "<a href=\"/grafana/\">Found</a>.\n\n", expLocation: "/grafana/"},
+		}
+
+		path := "/a/test-app/test"
+		for i, tc := range tcs {
+			t.Run(fmt.Sprintf("testcase %d", i), func(t *testing.T) {
+				ps := pluginstore.NewFakePluginStore(pluginstore.Plugin{
+					JSONData: plugins.JSONData{
+						ID: "test-app",
+						Includes: []*plugins.Includes{
+							{
+								Type: "page",
+								Role: tc.roleRequired,
+								Path: path,
+							},
+						},
+					},
+				})
+
+				middlewareScenario(t, t.Name(), func(t *testing.T, sc *scenarioContext) {
+					if tc.signedIn {
+						sc.withIdentity(&authn.Identity{
+							OrgRoles: map[int64]org.RoleType{
+								0: tc.role,
+							},
+						})
+					}
+					features := featuremgmt.WithFeatures()
+					logger := &logtest.Fake{}
+					ac := &actest.FakeAccessControl{}
+
+					sc.m.Get("/a/:id/*", RoleAppPluginAuthAndSignedIn(ac, ps, features, logger), func(c *contextmodel.ReqContext) {
```

Comment:
Nit. Wonder if using webtest package would be more suitable? Then you're testing the routes that's actually been registered in api.go. Or does it create an import cycle perhaps :thinking: Something for the future otherwise

Reply:
Aye an import cycle 😮‍💨